internal/credential/vault/service_list.go

```diff
@@ -0,0 +1,91 @@
+
+import (
+	"context"
+	"slices"
+	"time"
+
+	"github.com/hashicorp/boundary/internal/credential"
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/hashicorp/boundary/internal/errors"
+)
+
+func (s *LibraryService) List(ctx context.Context, credentialStoreId string, opts ...credential.Option) ([]credential.Library, error) {
+	const op = "vault.(*LibraryService).List"
+	opt, err := credential.GetOpts(opts...)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+	genericLibs, err := s.repo.ListCredentialLibraries(ctx, credentialStoreId, opts...)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+	sshCertLibs, err := s.repo.ListSSHCertificateCredentialLibraries(ctx, credentialStoreId, opts...)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+	libs := append(genericLibs, sshCertLibs...)
+	slices.SortFunc(libs, func(i, j credential.Library) int {
+		return i.GetUpdateTime().AsTime().Compare(j.GetUpdateTime().AsTime())
+	})
+	if opt.WithLimit == -1 {
+		// No limit, no truncation
+		return libs, nil
+	}
+	limit := s.repo.defaultLimit
+	if opt.WithLimit != 0 {
+		limit = opt.WithLimit
+	}
+	// Truncate slice to at most limit number of elements
+	if len(libs) > limit {
+		libs = libs[:limit]
+	}
+	return libs, nil
+}
+
+func (s *LibraryService) EstimatedCount(ctx context.Context) (int, error) {
+	const op = "vault.(*LibraryService).EstimatedCount"
+	numGenericLibs, err := s.repo.EstimatedLibraryCount(ctx)
+	if err != nil {
+		return 0, errors.Wrap(ctx, err, op)
+	}
+	numSSHCertLibs, err := s.repo.EstimatedSSHCertificateLibraryCount(ctx)
+	if err != nil {
+		return 0, errors.Wrap(ctx, err, op)
+	}
+	return numGenericLibs + numSSHCertLibs, nil
+}
+
+func (s *LibraryService) ListDeletedIds(ctx context.Context, since time.Time) ([]string, time.Time, error) {
+	const op = "vault.(*LibraryService).ListDeletedIds"
+	// Request and combine deleted ids from the DB for generic and ssh cert libraries.
+	// This statement here is the reason we need a struct for this. We need all the
+	// deleted libraries to be collated in a single transaction with a single
+	// transaction timestamp. This requires access to a db Reader, which isn't available
+	// to the handlers and can't be passed into this method. Therefore, a struct,
+	// constructed in the controller, is necessary.
+	var deletedIds []string
+	var transactionTimestamp time.Time
+	if _, err := s.writer.DoTx(ctx, db.StdRetryCnt, db.ExpBackoff{}, func(r db.Reader, w db.Writer) error {
+		deletedGenericIds, err := s.repo.ListDeletedLibraryIds(ctx, since, credential.WithReaderWriter(r, w))
+		if err != nil {
+			return err
+		}
+		deletedSSHCertIds, err := s.repo.ListDeletedSSHCertificateLibraryIds(ctx, since, credential.WithReaderWriter(r, w))
+		if err != nil {
+			return err
+		}
+		transactionTimestamp, err = r.Now(ctx)
+		if err != nil {
+			return err
+		}
+		deletedIds = append(deletedGenericIds, deletedSSHCertIds...)
+		return nil
+	}); err != nil {
```

Comment:
I went back and forth on whether to implement the transaction and timestamp here or in the credential service layer. I think from a design point of view it is more "correct" for this to be in the credential service layer, since in theory any future credential library implementation would need to have its ListDeletedIds called from that layer, and so that is where the transaction and timestamp would go. However, doing it this way lets us forgo the credential.LibraryService struct altogether, reducing clutter in the controller and the credential library application layer, so I went this route. It's not a big change to make if we add another credential library type in the future.

I'm happy to change it if this is a concern, of course 😄.